lib/mongo/collection/view/iterable.rb

```diff
@@ -42,12 +42,18 @@ def each
             # No need to store the cursor in the query cache if there is
             # already a cached cursor stored at this key.
             QueryCache.set(@cursor, cache_options) unless cached_cursor
-            range = limit || nil
+
+            # If a query with a limit is performed, the query cache will
+            # re-use results from an earlier query with the same or larger
+            # limit, and then impose the lower limit during iteration.
+            limit_for_cached_query = respond_to?(:limit) ? limit : nil
           end
 
           if block_given?
-            if range
-              @cursor.to_a[0...range].each do |doc|
+            # Ruby versions 2.5 and older do not support arr[0..nil] syntax, so
+            # this must be a separate conditional.
```

Comment:
Would slice work? https://ruby-doc.org/core-2.7.1/Array.html#method-i-slice

Reply:
It looks like `slice` only takes Integer arguments (get a `cannot convert nil to integer` exception).

Reply:
:+1: